include/drivers/pwm.h

```diff
@@ -186,6 +406,176 @@ static inline int pwm_pin_set_nsec(const struct device *dev, uint32_t pwm,
+ */
+static inline int pwm_pin_cycles_to_usec(const struct device *dev, uint32_t pwm,
+					 uint32_t cycles, uint64_t *usec)
+{
+	uint64_t cycles_per_sec;
+	uint64_t temp;
+
+	if (pwm_get_cycles_per_sec(dev, pwm, &cycles_per_sec) != 0) {
+		return -EIO;
+	}
+
+	if (u64_mul_overflow(cycles, (uint64_t)USEC_PER_SEC, &temp)) {
+		return -ERANGE;
+	}
+
+	*usec = temp / cycles_per_sec;
+
+	return 0;
+}
+
+/**
+ * @brief Convert from PWM cycles to nanoseconds.
+ *
+ * @param dev Pointer to the device structure for the driver instance.
+ * @param pwm PWM pin.
+ * @param cycles Cycles to be converted.
+ * @param nsec Pointer to the memory to store the calculated nsec.
+ *
+ * @retval 0 If successful.
+ * @retval -EIO If cycles per second cannot be determined.
+ * @retval -ERANGE If result is too large.
+ */
+static inline int pwm_pin_cycles_to_nsec(const struct device *dev, uint32_t pwm,
+					 uint32_t cycles, uint64_t *nsec)
+{
+	uint64_t cycles_per_sec;
+	uint64_t temp;
+
+	if (pwm_get_cycles_per_sec(dev, pwm, &cycles_per_sec) != 0) {
+		return -EIO;
+	}
+
+	if (u64_mul_overflow(cycles, (uint64_t)NSEC_PER_SEC, &temp)) {
+		return -ERANGE;
+	}
+
+	*nsec = temp / cycles_per_sec;
+
+	return 0;
+}
+
+/**
+ * @brief Capture a single PWM period/pulse width in microseconds for a single
+ *        PWM input.
+ *
+ * This API function wraps calls to @a pwm_pin_capture_cycles() and @a
+ * pwm_pin_cycles_to_usec() and passes the capture results to the caller. The
+ * function is blocking until either the PWM capture is completed or a timeout
+ * occurs.
+ *
+ * @param dev Pointer to the device structure for the driver instance.
+ * @param pwm PWM pin.
+ * @param flags PWM capture flags.
+ * @param period Pointer to the memory to store the captured PWM period width
+ *               (in usec).
+ * @param pulse Pointer to the memory to store the captured PWM pulse width (in
+ *              usec).
+ * @param timeout Waiting period for the capture to complete.
+ *
+ * @retval 0 If successful.
+ * @retval -EBUSY PWM capture already in progress.
+ * @retval -EAGAIN Waiting period timed out.
+ * @retval -EIO IO error while capturing.
+ * @retval -ERANGE If result is too large.
+ */
+
+static inline int pwm_pin_capture_usec(const struct device *dev, uint32_t pwm,
+				       pwm_flags_t flags,
+				       uint64_t *period,
+				       uint64_t *pulse,
+				       k_timeout_t timeout)
+{
+	uint32_t period_cycles;
+	uint32_t pulse_cycles;
+	int err;
+
+	err = pwm_pin_capture_cycles(dev, pwm, flags, &period_cycles,
+				     &pulse_cycles, timeout);
+	if (err) {
+		return err;
+	}
+
+	err = pwm_pin_cycles_to_usec(dev, pwm, period_cycles, period);
+	if (err) {
+		return err;
+	}
+
+	err = pwm_pin_cycles_to_usec(dev, pwm, pulse_cycles, pulse);
+	if (err) {
+		return err;
+	}
+
+	return 0;
+}
+
+/**
+ * @brief Capture a single PWM period/pulse width in nanoseconds for a single
+ *        PWM input.
+ *
+ * This API function wraps calls to @a pwm_pin_capture_cycles() and @a
```

Comment:
Do we need both nsec and usec implementations?   With 64 bit values nsec should cover all cases; the caller can divide if appropriate.

Though AFAICT this API already provides both, so I guess we have to continue with the duplication.

Reply:
Yeah, I added it both as a convenience and due to the API already supporting `cycles`, `nsec`, and `usec` variants for PWM generation.